src/java/org/apache/cassandra/db/compaction/CompactionController.java

```diff
@@ -275,10 +261,26 @@ public void close()
 
     /**
      * @param keyspace
-     * @return true if this node is currently receiving data as part of a range movement (e.g., bootstrap, decommission, move)
+     * @return true if this node may be streaming data for this keyspace, or if cassandra.never_purge_tombstones is set
+     * If we are streaming, tombstones should not be purged so that we don't pre-maturely purge those that exceed gc_grace_seconds.
+     * <p>
+     * Note that this node may receive data without streaming, e.g. during a repair.
+     */
+    private static boolean doNotPurgeTombstones(String keyspace)
+    {
+        return NEVER_PURGE_TOMBSTONES
+                || StorageService.instance.isRebuilding()
+                || pendingRangesExistForKeyspace(keyspace)
+                || StreamManager.instance.getTotalReceivingStreams() > 0;
```

Comment:
one worry I have is the new checks are going to limit all purging, rather than specifically for the keyspace that's streaming. I suppose we can start with this though and refactor if it becomes an issue

Reply:
we could start without the `StreamManager.instance.getTotalReceivingStreams() > 0`, which would keep the same keyspace-specific behavior for bootstraps ? and for rebuilding all the keyspaces are streamed so i think it'd be the same either way there

Reply:
you are able to rebuild one keyspace at a time, but yeah it's only tracked as a single boolean. for our purposes w/ how we do DC switches I think that's fine for now.

down to do your suggestion in this PR